### ISSUE TITLE
systemd-network crashes with SEGV

### ISSUE THREAD

**Issue by athei** [systemd/systemd#7760]
### Submission type

 Bug report

### systemd version the issue has been seen with
systemd **236**
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN default-hierarchy=hybrid

### Used distribution
**Arch Linux**
`Linux homebsd 4.14.8-1-ARCH #1 SMP PREEMPT Wed Dec 20 21:27:44 UTC 2017 x86_64 GNU/Linux`

### In case of bug report: Expected behaviour you didn't see
systemd-network should start and keep running without crashes.

### In case of bug report: Unexpected behaviour you saw
Systemd crashes immediately after discovering interfaces:
https://pastebin.com/76Yfnhtj (journalctl -b -u systemd-networkd)

```
PID: 14276 (systemd-network)
           UID: 193 (systemd-network)
           GID: 193 (systemd-network)
        Signal: 11 (SEGV)
     Timestamp: Fri 2017-12-29 17:40:03 CET (10min ago)
  Command Line: /usr/lib/systemd/systemd-networkd
    Executable: /usr/lib/systemd/systemd-networkd
 Control Group: /system.slice/systemd-networkd.service
          Unit: systemd-networkd.service
         Slice: system.slice
       Boot ID: d55348f881ad48df8817d3ff68a4e9dd
    Machine ID: f7a3181621d74eff99ca6e25677e8af3
      Hostname: homebsd
       Storage: /var/lib/systemd/coredump/core.systemd-network.193.d55348f881ad48df8817d3ff68a4e9dd.14276.1514565603000000.lz4
       Message: Process 14276 (systemd-network) of user 193 dumped core.
                
                Stack trace of thread 14276:
                #0  0x00007fda89cda496 __GI___strlen_sse2 (libc.so.6)
                #1  0x0000559791a62fd9 n/a (systemd-networkd)
                #2  0x00007fda8990fc28 n/a (libsystemd-shared-236.so)
                #3  0x00007fda89910ad8 set_put (libsystemd-shared-236.so)
                #4  0x0000559791a6e19d n/a (systemd-networkd)
                #5  0x0000559791a57d6c n/a (systemd-networkd)
                #6  0x0000559791a52474 n/a (systemd-networkd)
                #7  0x00007fda89c64f4a __libc_start_main (libc.so.6)
                #8  0x0000559791a5290a n/a (systemd-networkd)
```

http://www55.zippyshare.com/v/Lpl5rnOB/file.html (Coredump)

### In case of bug report: Steps to reproduce the problem

I am not sure. I do not do anything special here. It started happening around the time I upgraded from v235 to v236. I do **not** have any files in /etc/systemd/network.

**Comment by yaka-es** [comment]
I have ran into this bug too. After systemd upgrade (235.38-1 -> 236.0-2) systemd-networkd crashes with the following messages:
```
Dec 30 06:47:42 gw systemd[1]: Starting Network Service...
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/vlan3.netdev:10: Unknown section 'VLAN'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/vlan2.netdev:10: Unknown section 'VLAN'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/vlan18.netdev:10: Unknown section 'VLAN'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/vlan17.netdev:10: Unknown section 'VLAN'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/vlan16.netdev:10: Unknown section 'VLAN'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/vlan1.netdev:10: Unknown section 'VLAN'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/br3.netdev:10: Unknown section 'Bridge'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/br2.netdev:10: Unknown section 'Bridge'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/br1.netdev:10: Unknown section 'Bridge'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/br0.netdev:10: Unknown section 'Bridge'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: /etc/systemd/network/br-ext.netdev:10: Unknown section 'Bridge'. Ignoring.
Dec 30 06:47:43 gw systemd-networkd[10887]: br-ext-dummy: netdev ready
Dec 30 06:47:43 gw systemd-networkd[10887]: br-ext: netdev ready
Dec 30 06:47:43 gw systemd-networkd[10887]: br0: netdev ready
Dec 30 06:47:43 gw systemd-networkd[10887]: br1: netdev ready
Dec 30 06:47:43 gw systemd-networkd[10887]: br2: netdev ready
Dec 30 06:47:43 gw systemd-networkd[10887]: br3: netdev ready
Dec 30 06:47:43 gw systemd-networkd[10887]: tap-obs: Gained IPv6LL
Dec 30 06:47:43 gw systemd-networkd[10887]: tap-remote: Gained IPv6LL
Dec 30 06:47:43 gw systemd-networkd[10887]: vb-gw-old: Gained IPv6LL
Dec 30 06:47:43 gw systemd[1]: systemd-networkd.service: Main process exited, code=dumped, status=11/SEGV
Dec 30 06:47:43 gw systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
Dec 30 06:47:43 gw systemd[1]: Failed to start Network Service.
Dec 30 06:47:43 gw systemd[1]: systemd-networkd.service: Service has no hold-off time, scheduling restart.
Dec 30 06:47:43 gw systemd[1]: systemd-networkd.service: Scheduled restart job, restart counter is at 1.
Dec 30 06:47:43 gw systemd[1]: Stopped Network Service.
```
In my case there are several .network and .netdev files in /etc/systemd/network. For example:
```
# etc/systemd/network/br0.netdev

[NetDev]
Name=br0
Kind=bridge

[Bridge]
STP=on
HelloTimeSec=1
ForwardDelaySec=2
```
```
# etc/systemd/network/br0.network

[Match]
Name=br0

[Network]
DHCP=ipv4
LLMNR=no
LLDP=yes
EmitLLDP=customer-bridge
IPForward=yes

VLAN=vlan1
VLAN=vlan2
VLAN=vlan3
VLAN=vlan16
VLAN=vlan17
VLAN=vlan18
```
```
# etc/systemd/network/vlan3.netdev

[NetDev]
Name=vlan3
Kind=vlan

[VLAN]
Id=3
```
```
# etc/systemd/network/vlan3.network

[Match]
Name=vlan3

[Network]
LLMNR=no
LLDP=no
EmitLLDP=no
Bridge=br3
```

**Comment by jwwalton** [comment]
rule being present or added (ip rule add from 192.168.178.244 lookup 3) is enough to make it crash in various conditions

```
(gdb) bt
#0  0x00007f34fc674ed7 in __strlen_avx2 () from /usr/lib/libc.so.6
#1  0x000055fad93f9244 in routing_policy_rule_hash_func.lto_priv.138 (b=0x55fadb46c640, state=0x7ffc40dfa4b0) at ../systemd-stable/src/network/networkd-routing-policy-rule.c:99
#2  0x00007f34fc1eaab7 in base_bucket_hash.lto_priv.173 (h=h@entry=0x55fadb481790, p=p@entry=0x55fadb46c640) at ../systemd-stable/src/basic/hashmap.c:346
#3  0x00007f34fc1ebcc8 in set_put (s=<optimized out>, key=<optimized out>) at ../systemd-stable/src/basic/hashmap.c:1247
#4  0x000055fad94041cc in routing_policy_rule_add_internal.lto_priv.148 (rules=0x55fadb45fdf0, family=2, from=0x7ffc40dfa620, from_prefixlen=<optimized out>, to=0x7ffc40dfa630, to_prefixlen=<optimized out>, 
    tos=0 '\000', fwmark=0, table=3, iif=0x7ffc40dfa690 "\260\367E\333\372U", oif=0x409f69 <error: Cannot access memory at address 0x409f69>, ret=0x7ffc40dfa618)
    at ../systemd-stable/src/network/networkd-routing-policy-rule.c:276
#5  0x000055fad93ee247 in routing_policy_rule_add_foreign (ret=0x7ffc40dfa618, oif=<optimized out>, iif=<optimized out>, table=<optimized out>, fwmark=<optimized out>, tos=<optimized out>, 
    to_prefixlen=<optimized out>, to=0x7ffc40dfa630, from_prefixlen=<optimized out>, from=0x7ffc40dfa620, family=<optimized out>, m=0x55fadb45fcc0) at ../systemd-stable/src/network/networkd-routing-policy-rule.c:316
#6  manager_rtnl_process_rule (rtnl=<optimized out>, message=<optimized out>, userdata=0x55fadb45fcc0) at ../systemd-stable/src/network/networkd-manager.c:821
#7  0x00007f34fc18ac76 in process_match (m=0x55fadb473410, rtnl=0x55fadb4615a0) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:373
#8  process_running (ret=0x0, rtnl=0x55fadb4615a0) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:403
#9  sd_netlink_process (rtnl=0x55fadb4615a0, ret=ret@entry=0x0) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:437
#10 0x00007f34fc18aeb3 in io_callback (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=<optimized out>) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:731
#11 0x00007f34fc207800 in source_dispatch (s=s@entry=0x55fadb461660) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2270
#12 0x00007f34fc207a3a in sd_event_dispatch (e=e@entry=0x55fadb45f510) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2627
#13 0x00007f34fc207bb7 in sd_event_run (e=e@entry=0x55fadb45f510, timeout=timeout@entry=18446744073709551615) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2686
#14 0x00007f34fc207d9b in sd_event_loop (e=<optimized out>, e=<optimized out>) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2706
#15 0x000055fad93e8dde in main (argc=<optimized out>, argv=<optimized out>) at ../systemd-stable/src/network/networkd.c:154



(gdb) thread apply all bt full

Thread 1 (Thread 0x7f34fccdc8c0 (LWP 8533)):
#0  0x00007f34fc674ed7 in __strlen_avx2 () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x000055fad93f9244 in routing_policy_rule_hash_func.lto_priv.138 (b=0x55fadb46c640, state=0x7ffc40dfa4b0) at ../systemd-stable/src/network/networkd-routing-policy-rule.c:99
        rule = 0x55fadb46c640
        __PRETTY_FUNCTION__ = "routing_policy_rule_hash_func"
#2  0x00007f34fc1eaab7 in base_bucket_hash.lto_priv.173 (h=h@entry=0x55fadb481790, p=p@entry=0x55fadb46c640) at ../systemd-stable/src/basic/hashmap.c:346
        state = {v0 = 10517170141054663664, v1 = 6169725426090971619, v2 = 5480016767541823477, v3 = 4376858984443840236, padding = 549692956582, inlen = 29}
        hash = <optimized out>
#3  0x00007f34fc1ebcc8 in set_put (s=<optimized out>, key=<optimized out>) at ../systemd-stable/src/basic/hashmap.c:1247
        swap = {e = {{p = {b = {key = 0x2}, value = 0x7f34fc1e9eeb <hashmap_base_new.lto_priv.176+123>}, iterate_next = 3678994720, iterate_previous = 22010}, {p = {b = {key = 0x55fadb45fdf0}, value = 0x0}, iterate_next = 0, iterate_previous = 0}}}
        e = <optimized out>
        hash = <optimized out>
        idx = <optimized out>
        __PRETTY_FUNCTION__ = "set_put"
#4  0x000055fad94041cc in routing_policy_rule_add_internal.lto_priv.148 (rules=0x55fadb45fdf0, family=2, from=0x7ffc40dfa620, from_prefixlen=<optimized out>, to=0x7ffc40dfa630, to_prefixlen=<optimized out>, tos=0 '\000', fwmark=0, table=3, iif=0x7ffc40dfa690 "\260\367E\333\372U", 
    oif=0x409f69 <error: Cannot access memory at address 0x409f69>, ret=0x7ffc40dfa618) at ../systemd-stable/src/network/networkd-routing-policy-rule.c:276
        rule = 0x55fadb46c640
        r = <optimized out>
        __PRETTY_FUNCTION__ = "routing_policy_rule_add_internal"
#5  0x000055fad93ee247 in routing_policy_rule_add_foreign (ret=0x7ffc40dfa618, oif=<optimized out>, iif=<optimized out>, table=<optimized out>, fwmark=<optimized out>, tos=<optimized out>, to_prefixlen=<optimized out>, to=0x7ffc40dfa630, from_prefixlen=<optimized out>, from=0x7ffc40dfa620, family=<optimized out>, 
    m=0x55fadb45fcc0) at ../systemd-stable/src/network/networkd-routing-policy-rule.c:316
No locals.
#6  manager_rtnl_process_rule (rtnl=<optimized out>, message=<optimized out>, userdata=0x55fadb45fcc0) at ../systemd-stable/src/network/networkd-manager.c:821
        tos = 0 '\000'
        to_prefixlen = 0 '\000'
        from_prefixlen = 32 ' '
        rule = 0x0
        to = {in = {s_addr = 3678853040}, in6 = {__in6_u = {__u6_addr8 = "\260\327F\333\372U\000\000\330\023\000\000\000\000\000", __u6_addr16 = {55216, 56134, 22010, 0, 5080, 0, 0, 0}, __u6_addr32 = {3678853040, 22010, 5080, 0}}}}
        from = {in = {s_addr = 4105349312}, in6 = {__in6_u = {__u6_addr8 = "\300\250\262\364\064\177\000\000\020\064G\333\372U\000", __u6_addr16 = {43200, 62642, 32564, 0, 13328, 56135, 22010, 0}, __u6_addr32 = {4105349312, 32564, 3678876688, 22010}}}}
        fwmark = 0
        table = 3
        m = 0x55fadb45fcc0
        iif = 0x7ffc40dfa690 "\260\367E\333\372U"
        oif = 0x409f69 <error: Cannot access memory at address 0x409f69>
        type = 32
        family = 2
        r = <optimized out>
        __PRETTY_FUNCTION__ = "manager_rtnl_process_rule"
        __func__ = "manager_rtnl_process_rule"
#7  0x00007f34fc18ac76 in process_match (m=0x55fadb473410, rtnl=0x55fadb4615a0) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:373
        c = 0x55fadb461d50
        type = 32
        r = <optimized out>
#8  process_running (ret=0x0, rtnl=0x55fadb4615a0) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:403
        m = 0x55fadb473410
        r = <optimized out>
#9  sd_netlink_process (rtnl=0x55fadb4615a0, ret=ret@entry=0x0) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:437
        _dont_destroy_rtnl = 0x55fadb4615a0
        __PRETTY_FUNCTION__ = "sd_netlink_process"
#10 0x00007f34fc18aeb3 in io_callback (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=<optimized out>) at ../systemd-stable/src/libsystemd/sd-netlink/sd-netlink.c:731
        rtnl = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "io_callback"
#11 0x00007f34fc207800 in source_dispatch (s=s@entry=0x55fadb461660) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2270
        saved_type = SOURCE_IO
        r = <optimized out>
        __PRETTY_FUNCTION__ = "source_dispatch"
        __func__ = "source_dispatch"
#12 0x00007f34fc207a3a in sd_event_dispatch (e=e@entry=0x55fadb45f510) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2627
        p = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "sd_event_dispatch"
#13 0x00007f34fc207bb7 in sd_event_run (e=e@entry=0x55fadb45f510, timeout=timeout@entry=18446744073709551615) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2686
        r = 1
        __PRETTY_FUNCTION__ = "sd_event_run"
#14 0x00007f34fc207d9b in sd_event_loop (e=<optimized out>, e=<optimized out>) at ../systemd-stable/src/libsystemd/sd-event/sd-event.c:2706
        r = <optimized out>
        e = 0x55fadb45f510
        __PRETTY_FUNCTION__ = "sd_event_loop"
#15 0x000055fad93e8dde in main (argc=<optimized out>, argv=<optimized out>) at ../systemd-stable/src/network/networkd.c:154
        event = 0x55fadb45f510
        m = 0x55fadb45fcc0
        user = 0x55fad94314c5 "systemd-network"
        uid = 193
        gid = 193
        r = <optimized out>
        __func__ = "main"
        __PRETTY_FUNCTION__ = "main"


```

**Comment by yuwata** [comment]
Could you try #7712? This seems duplicate of #7704 or #7708.

**Comment by amishmm** [comment]
I have run in to same / similar issue too.

Using archlinux and systemd 236.0-2

I also have multiple interface and I also run ip route / rule add externally.

> Jan 03 13:00:02 sp systemd-networkd[3029]: eth2: Gained IPv6LL
> Jan 03 13:00:02 sp systemd-networkd[3029]: eth0: Gained IPv6LL
> Jan 03 13:00:02 sp systemd-networkd[3029]: eth1: Gained IPv6LL
> Jan 03 13:00:02 sp systemd[1]: Started Process Core Dump (PID 3030/UID 0).
> Jan 03 13:00:02 sp systemd[1]: systemd-networkd.service: Main process exited, code=dumped, status=11/SEGV
> Jan 03 13:00:02 sp systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
> Jan 03 13:00:02 sp systemd[1]: Failed to start Network Service.
> Jan 03 13:00:02 sp systemd[1]: systemd-networkd.service: Service has no hold-off time, scheduling restart.
> Jan 03 13:00:02 sp systemd[1]: systemd-networkd.service: Scheduled restart job, restart counter is at 5.
> Jan 03 13:00:02 sp systemd[1]: Stopped Network Service.
> Jan 03 13:00:02 sp systemd[1]: systemd-networkd.service: Start request repeated too quickly.
> Jan 03 13:00:02 sp systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
> Jan 03 13:00:02 sp systemd[1]: Failed to start Network Service.
> Jan 03 13:00:02 sp systemd-coredump[3017]: Process 3014 (systemd-network) of user 193 dumped core.
>                                                     
>                                                     Stack trace of thread 3014:
>                                                     #0  0x00007f76efe36461 __strlen_avx2 (libc.so.6)
>                                                     #1  0x000056307c476fd9 n/a (systemd-networkd)
>                                                     #2  0x00007f76ef9aac28 n/a (libsystemd-shared-236.so)
>                                                     #3  0x00007f76ef9abad8 set_put (libsystemd-shared-236.so)
>                                                     #4  0x000056307c48219d n/a (systemd-networkd)
>                                                     #5  0x000056307c46bd6c n/a (systemd-networkd)
>                                                     #6  0x000056307c466474 n/a (systemd-networkd)
>                                                     #7  0x00007f76efcfff4a __libc_start_main (libc.so.6)
>                                                     #8  0x000056307c46690a n/a (systemd-networkd)

**Comment by poettering** [comment]
I am pretty sure this has been fixed by #7712. Let's close this here hence. if the problem persists even with versions including that, please reopen.
